go-metadata/istio.go

```diff
@@ -0,0 +1,141 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istiometadata
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net"
+	"net/http"
+	"net/url"
+	"os"
+	"time"
+)
+
+// ConnectionMetadata returns the metadata about a connection
+type ConnectionMetadata struct {
+	// Identity provides the identity of the peer.
+	// Example: spiffe://cluster.local/ns/a/sa/b.
+	Identity string `json:"identity"`
```

Comment:
Can we open a tracking issue to capture the other information about the identity that we would like to expose here

Over time I would expect to see us add either a field or a new API that provided details from the public key of the peer

I think the interesting question is what are we trying to prove to the caller about the veracity of the identity we just provided. Options include:

1. *How can the caller know that the MDS is reliably acting on its behalf*. Clearly theres a certain amount of infrastructural trust seeing as this is coming from the 'network' as perceived by the app. One option here if for the MDS to offer a "who am I" API. This is a bit like SDS but now its an implicit part of the infrastructure. Response to the call should either be the current public cert or some shorthand of it. Variant, can prove that the infra has the private key which is shielded from the app by signing some random bytes passed to the API which can then be verified with the public key. Likely want to deliver a shorthand in addition to x509 to shield users from parsing its gory details.

2. *How can the caller know that the infrastructure actually performed a secure handshake with the identified peer?* This is harder and of debateable value. If we provide proof that infrastructure already has the public and private keys as described in 1 above then the infra is already capable of impersonating the application for whatever purposes it sees fit. We could do something like a 'signed ping'
          
         Client generates random bytes and encrypts with public key of peer (opaque to local ztunnel)
         Client app calls 'secure ping' passing the encrypted bytes
         local ztunnel simply forwards encrypted bytes to the identified peer
         peer ztunnel decrypts random byes with private key and returns their signature
         local ztunnel forwars message signature to client
         Client compares message signature to locally computed hash

This proves that there is something between the client and the peer that has access to the private key associated with the identity. The client doesn't know who or what that is necessarily

Reply:
```
Over time I would expect to see us add either a field or a new API that provided details from the public key of the peer
```

ideally one that follows XFCC semantics, and chains certs of all workloads in the pathway, including the ztunnel instance.

Reply:
XFCC is a pretty horrible UX - you need a parser library, parse certs, etc. It may be useful for some super advanced users - but if they are willing to take all those dependencies they may be better off just using proxyless and having end to end mTLS.

All platforms I know ( not only google) have a basic, simple MDS server that returns metadata and tokens in 
a very simple way. 

I'm not saying it is an invalid use case to want fancy attestation - and it is clear that cloud vendors don't 
provide this right now, so it's not a bad idea if someone implemented it - but we should not mix it with
the very simple use case for this PR.